### PR TITLE
Removing empty usage on decoding

### PR DESCRIPTION
**Pull request by nyamsprod**
As a follow up to #33 I'm also removing `empty` usage on decompiling the Problem object to remove the following bug

```php
<?php

echo ApiProblem::fromJson('{"type":"0"}')->asJson();
// expected {"type":"0"} but got {"type":"about:blank"} instead
```

### CODE REVIEW

**Comment by nyamsprod** [comment]
I could replace the use of `filter_var` with `intval` and `strval` if you want to rely on functions only present in cores. What do you think ?

File: src/ApiProblem.php [hunk 1]
```diff
@@ -135,7 +135,7 @@ class ApiProblem implements \ArrayAccess, \JsonSerializable
      */
     public static function fromJson(string $json) : self
     {
-        if (empty($json)) {
+        if ($json === '') {
```
Comment (Crell) [inline src/ApiProblem.php:138]:
```suggestion
        if (trim($json) === '') {
```

An all-whitespace string wouldn't be valid JSON either, I believe.

File: src/ApiProblem.php [hunk 2]
```diff
@@ -215,20 +215,28 @@ protected static function decompile(array $parsed) : self
...
         $problem = new static();
 
-
-        if (!empty($parsed['title'])) {
+        // Skip empty string or missing values.
+        // The string or integer 0, however, are allowed.  PHP makes
+        // this ugly. The check on string handles XML decompile which
+        // may return an empty array.
+        $emptyValues = ['', []];
+        if (isset($parsed['title']) && !in_array($parsed['title'], $emptyValues, true)) {
```
Comment (Crell) [inline src/ApiProblem.php:223]: This check seems like something that should be factored out to a utility method.  It's basically a more forgiving `is_empty()`.  It's used repeatedly, though, so let's make it its own method and move the comment describing it there.

Suggested name: `validValue(string $str): bool` (so it ends up reading `if(static::validValue($parsed['title']){ ... }`

File: src/ApiProblem.php [hunk 3]
```diff
@@ -215,20 +215,28 @@ protected static function decompile(array $parsed) : self
...
         }
-        if (!empty($parsed['type'])) {
+
+        if (isset($parsed['type']) && !in_array($parsed['type'], $emptyValues, true)) {
             $problem->setType($parsed['type']);
         }
-        if (!empty($parsed['status'])) {
-            $problem->setStatus((int) $parsed['status']);
+
+        if (isset($parsed['status']) && ($status = filter_var($parsed['status'], FILTER_VALIDATE_INT)) !== false) {
```
Comment (Crell) [inline src/ApiProblem.php:231]: Wouldn't `is_numeric()` and `intval()` work better here?  I suppose that would end up truncating a float, but if you send a float in here I wouldn't expect anything logical to happen.

(At some point I should bump this library to PHP 7.4 so I can type the properties.  Not right now, though.)

File: src/ApiProblem.php (hunk 1 above)
Comment (nyamsprod) [reply src/ApiProblem.php:138]: Indeed you are correct. I'll update this code

File: src/ApiProblem.php (hunk 3 above)
Comment (nyamsprod) [reply src/ApiProblem.php:231]: I will use my alternate solution with `intval` and `strval` which will also reject `float`.

**Comment by nyamsprod** [comment]
I've updated the PR by adding 2 static protected methods `isValidString` and `parseInt` using `intval` and `strval`

**Comment by Crell** [comment]
Dur.  I thought I told Scrutinizer to buzz off.

**Comment by Crell** [comment]
I really hate loose typing... :smile: 

Any other updates you have in mind?  I don't want to have too many more tags in rapid succession.

**Comment by nyamsprod** [comment]
At the moment no those are just issues I stumble on during CI/CD tests. From my part it should be fine for now. If you can wait 1 week I'll come back to you with a definitive go ahead from my part.

**Comment by Crell** [comment]
:+1:

**Comment by nyamsprod** [comment]
@Crell all is good you can proceed with the release 😉👌🏾

**Pull request merged**